packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/SurroundingPairScopeHandler/SurroundingPairScopeHandler.ts

```diff
@@ -0,0 +1,121 @@
+import { getSurroundingPairOccurrences } from "./getSurroundingPairOccurrences";
+import { ide } from "../../../../singletons/ide.singleton";
+import { createTargetScope } from "./createTargetScope";
+import { stronglyContains } from "./stronglyContains";
+
+export class SurroundingPairScopeHandler extends BaseScopeHandler {
+  public readonly iterationScopeType;
+  protected isHierarchical = true;
+
+  constructor(
+    private languageDefinitions: LanguageDefinitions,
+    public readonly scopeType: SurroundingPairScopeType,
+    private languageId: string,
+  ) {
+    super();
+    const iterationScopeType: ScopeType = {
+      type: "oneOf",
+      scopeTypes: [this.scopeType, { type: "document" }],
+    };
+    this.iterationScopeType = iterationScopeType;
+  }
+
+  *generateScopeCandidates(
+    editor: TextEditor,
+    position: Position,
+    direction: Direction,
+    hints: ScopeIteratorRequirements,
+  ): Iterable<TargetScope> {
+    const individualDelimiters = getIndividualDelimiters(
+      this.scopeType.delimiter,
+      this.languageId,
+    );
+    const delimiterRegex = getDelimiterRegex(individualDelimiters);
+
+    if (this.scopeType.forceDirection != null) {
+      //  DEPRECATED @ 2024-07-01
+      void showError(
+        ide().messages,
+        "deprecatedForceDirection",
+        "forceDirection not supported. Use 'next pair' or 'previous pair' instead",
+      );
+      return;
+    }
+
+    const delimiterOccurrences = getDelimiterOccurrences(
+      this.languageDefinitions.get(this.languageId),
+      editor.document,
+      individualDelimiters,
+      delimiterRegex,
+    );
+
+    const surroundingPairs = getSurroundingPairOccurrences(
+      individualDelimiters,
+      delimiterOccurrences,
+    );
+
+    const { requireStrongContainment } = this.scopeType;
+
+    const scopes: TargetScope[] = [];
+
+    surroundingPairs.forEach((pair, i) => {
+      if (direction === "forward") {
+        if (pair.rightEnd.isBefore(position)) {
+          return;
+        }
+        // In the case of `(()|)` don't yield the pair to the left
+        if (pair.rightEnd.isEqual(position) && hints.skipAncestorScopes) {
+          const nextPair = surroundingPairs[i + 1];
+          if (nextPair != null && nextPair.rightStart.isEqual(position)) {
+            return;
+          }
+        }
+      } else {
+        if (pair.leftStart.isAfter(position)) {
+          return;
+        }
+      }
+      if (requireStrongContainment && !stronglyContains(position, pair)) {
+        return;
+      }
+      scopes.push(createTargetScope(editor, pair));
+    });
+
+    scopes.sort((a, b) => compareTargetScopes(direction, position, a, b));
+
+    for (const scope of scopes) {
+      yield scope;
+    }
+  }
+
+  isPreferredOver(
+    scopeA: TargetScope,
+    scopeB: TargetScope,
+    position: Position,
+  ): boolean | undefined {
+    // In the case of `( [ )| ]` we prefer the one we're actually touching
```

Comment:
Is that not the case for other scopes? Surprised this was necessary

Reply:
Do we have any other examples of overlapping scopes?

Reply:
No we don't support overlapping scopes. Are you suggesting we should support them? Guessing there was a failing test?

Reply:
`(  [  )|  ]` is an overlapping scope which we have tests for. That is why I need this work around.